### PR TITLE
adding representativePeer struct

### PR DESCRIPTION
**Pull request by shireenf-ibm**
issue #236

subtask:

>    *  - [ ] implement RepresentativePeer interface/struct

### CODE REVIEW

File: pkg/netpol/eval/internal/k8s/peer.go [hunk 1]
```diff
@@ -58,6 +58,13 @@ type WorkloadPeer struct {
 	Pod *Pod
 }
 
+// RepresentativePeer implements eval.Peer interface
+// a representative peer is a peer inferred from a policy rule not a parsed pod/deployment object
```
Comment (adisos) [inline pkg/netpol/eval/internal/k8s/peer.go:62]:
```suggestion
// a representative peer is a peer inferred from a policy rule (selector) not a parsed pod/deployment object
// and is used to represent a potential pod/ns entity in the cluster (that does not exist on the input resources)
// but may have enabled connectivity to input resources (pods/deployments) based on input network policies. 
```

**Review by adisos — CHANGES_REQUESTED** [review]

File: pkg/netpol/connlist/connlist.go [hunk 2]
```diff
@@ -514,6 +509,10 @@ func (ca *ConnlistAnalyzer) getConnectionsBetweenPeers(pe *eval.PolicyEngine, pe
 			if !ca.includePairOfWorkloads(srcPeer, dstPeer) {
 				continue
 			}
+			// when exposure-analysis, skip conns between fake pods or ip-peer and fake pods
+			if ca.exposureAnalysis && ca.hasFakePodsAndIPs(pe, srcPeer, dstPeer) {
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (shireenf-ibm) [reply pkg/netpol/connlist/connlist.go:515]: `hasFakePodsAndIPs` requires also the "pe" (*PolicyEngine) as a param; 
 so preferred to add it here than to change  `includePairOfWorkloads `

File: pkg/netpol/connlist/exposure_map.go [hunk 3]
```diff
@@ -122,10 +122,15 @@ func (ex exposureMap) addConnToExposureMap(pe *eval.PolicyEngine, allowedConnect
 	if _, ok := ex[peer]; !ok {
 		ex.initiatePeerEntry(peer)
 	}
+
+	nsLabels, err := pe.GetPeerNsLabels(inferredPeer)
```
[the remark this answers is not in the document]
Comment (shireenf-ibm) [reply pkg/netpol/connlist/exposure_map.go]: done

File: pkg/netpol/connlist/exposure_map.go [hunk 4]
```diff
@@ -122,10 +122,15 @@ func (ex exposureMap) addConnToExposureMap(pe *eval.PolicyEngine, allowedConnect
...
+	if err != nil {
+		return err
+	}
 	// store connection data
 	expData := &xgressExposure{
 		exposedToEntireCluster: false,
-		namespaceLabels:        pe.GetPeerNsLabels(inferredPeer),
+		namespaceLabels:        nsLabels,
```
[the remark this answers is not in the document]
Comment (shireenf-ibm) [reply pkg/netpol/connlist/exposure_map.go:133]: done

File: pkg/netpol/connlist/connlist.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/netpol/connlist/connlist.go:515]: I think it is also related to the same functionality as of the `includePairOfWorkloads` function, so better add it there, including the additional required param.

[the remark this answers is not in the document]
Comment (shireenf-ibm) [reply pkg/netpol/connlist/connlist.go]: done

File: pkg/netpol/eval/resources.go [hunk 5]
```diff
@@ -442,6 +451,7 @@ func (pe *PolicyEngine) createPodOwnersMap() (map[string]Peer, error) {
 
 // GetPeersList returns a slice of peers from all PolicyEngine resources
 // get peers in level of workloads (pod owners) of type WorkloadPeer, and ip-blocks
+// and representative peers if exposure-analysis is on
```
[the remark this answers is not in the document]
Comment (shireenf-ibm) [reply pkg/netpol/eval/resources.go]: done

File: pkg/netpol/eval/internal/k8s/peer.go [hunk 6]
```diff
@@ -58,6 +58,13 @@ type WorkloadPeer struct {
...
+type RepresentativePeer struct {
+	Pod                      *Pod
```
[the remark this answers is not in the document]
Comment (shireenf-ibm) [reply pkg/netpol/eval/internal/k8s/peer.go:73]: done (the comment may be updated when supporting rules with podSelector)

**Review by adisos — APPROVED** [review]

**Pull request merged**